src/linear_algebra/tensor_algebra.lean

```diff
@@ -0,0 +1,378 @@
+/-
+Copyright (c) 2020 Adam Topaz. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Adam Topaz.
+-/
+
+import ring_theory.algebra
+import linear_algebra
+
+/-!
+# Tensor Algebras
+
+Given a commutative ring R, and an R-module M, we construct the tensor algebra of M.
+This is the free R-algebra generated by the module M.
+
+## Notation
+
+1. `tensor_algebra R M` is the tensor algebra itself.
+  It is endowed with an R-algebra structure.
+2. `univ R M` is the R-linear map `M → tensor_algebra R M`.
+3. Given a linear map `f : M → A` to an R-algebra A,
+  `lift R M f` is the lift of `f` to an R-algebra morphism
+  `tensor_algebra R M → A`.
+
+## Theorems
+
+1. `univ_comp_lift` states that the composition
+  `(lift R M f) ∘ (univ R M)` is identical to `f`.
+2. `lift_unique` states that whenever an R-algebra morphism
+  `g : tensor_algebra R M → A` is given whose composision with `univ R M` is `f`,
+  then one has `g = lift R M f`.
+-/
+
+universes u1 u2 u3
+
+variables (R : Type u1) [comm_ring R]
+variables (M : Type u2) [add_comm_group M] [module R M]
+
+namespace tensor_algebra
+
+-- This inductive type is used to express representatives
+-- of the tensor algebra.
+@[nolint doc_blame]
+inductive pre
+| of : M → pre
+| zero : pre
+| one : pre
+| mul : pre → pre → pre
+| add : pre → pre → pre
+| smul : R → pre → pre
+
+
+namespace pre
+
+instance : inhabited (pre R M) := ⟨pre.zero⟩
+
+-- These instances are only used to simplify the notation.
+@[nolint doc_blame]
+def has_coe : has_coe M (pre R M) := ⟨of⟩
+@[nolint doc_blame]
+def has_mul : has_mul (pre R M) := ⟨mul⟩
+@[nolint doc_blame]
+def has_add : has_add (pre R M) := ⟨add⟩
+@[nolint doc_blame]
+def has_zero : has_zero (pre R M) := ⟨zero⟩
+@[nolint doc_blame]
+def has_one : has_one (pre R M) := ⟨one⟩
+@[nolint doc_blame]
+def has_scalar : has_scalar R (pre R M) := ⟨smul⟩
+@[nolint doc_blame]
+def has_neg : has_neg (pre R M) := ⟨smul (-1 : R)⟩
+
+end pre
+
+local attribute [instance] pre.has_coe
+local attribute [instance] pre.has_mul
+local attribute [instance] pre.has_add
+local attribute [instance] pre.has_zero
+local attribute [instance] pre.has_one
+local attribute [instance] pre.has_scalar
+local attribute [instance] pre.has_neg
+
+@[nolint doc_blame]
+def lift_fun {A : Type u3} [ring A] [algebra R A] (f : M →ₗ[R] A) : pre R M → A :=
+  λ t, pre.rec_on t f 0 1 (λ _ _, (*)) (λ _ _, (+)) (λ x _ a, x • a)
+
+@[nolint doc_blame]
+inductive rel : (pre R M) → (pre R M) → Prop
+-- force of to be linear
+| add_lin {a b : M} : rel ↑(a + b) (↑a + ↑b)
+| smul_lin {r : R} {a : M} : rel ↑(r • a) (r • ↑a)
+| zero_lin : rel ↑(0 : M) 0
```

Comment:
It would be neat if there was some way to reuse `is_linear_map` or `linear_map` here.

On that note, `zero_lin` is unecessary as you can get it from simplifying `@smul_lin 0 _` I think

Reply:
I don't think you can get by with only `smul_lin` since the `0` from `M` a priori has nothing to do with `pre.zero`.

It's possible there's some way to combine this with some of the other relations, but I feel like a small bit of redundancy is fine if it makes the construction more understandable.

Reply:
Could you eliminate `pre.zero` and replace it with `pre.of 0`? or does that make things harder below?

Reply:
That's possible. The thing is, I can imagine in the future having a more general construction which constructs rings given certain input data/relations, and this being a special case. This is mainly why I wanted to explicitly include all of the usual constructors for an algebra in pre, so that it would be easier to generalize later.

Reply:
On second thought, I think it would make sense to change the constructors to something like this:
```lean
inductive pre
| of : M → pre
| const : R → pre
| mul : pre → pre → pre
| add : pre → pre → pre
```
I'm making this change now.

Reply:
`of_scalar` might be a better name than `const`

Reply:
That construction looks harder to prove `has_scalar` in though.

Reply:
Not really. It's just multiplication by something of the form `of_scalar r`